### PR TITLE
Improve JSON format for categorical features.

### PR DESCRIPTION
**Pull request by trivialfis**
* This PR reduces the saved categories to a single vector instead of vector per-node.

### CODE REVIEW

**Comment by hcho3** [comment]
@trivialfis Do you have performance numbers? How much does this PR improve serialization performance?

File: src/tree/tree_model.cc [hunk 1]
```diff
@@ -772,6 +776,92 @@ void RegTree::Save(dmlc::Stream* fo) const {
...
+      categories_nodes.emplace_back(i);
+      auto begin = categories.size();
+      categories_segments.emplace_back(static_cast<Integer::Int>(begin));
+      auto segment = split_categories_segments_[i];
+      auto node_categories =
+          this->GetSplitCategories().subspan(segment.beg, segment.size);
+      common::KCatBitField const cat_bits(node_categories);
+      for (size_t i = 0; i < cat_bits.Size(); ++i) {
+        if (cat_bits.Check(i)) {
+          categories.emplace_back(static_cast<Integer::Int>(i));
```
Comment (hcho3) [inline src/tree/tree_model.cc:851]: @trivialfis What's the performance implication of saving individual categories? Is it better than saving bitmaps directly?

Comment (trivialfis) [reply src/tree/tree_model.cc:851]: No, it's not.  If saving bitmap is preferred I can make the change.

Comment (hcho3) [reply src/tree/tree_model.cc:851]: No, if saving individual categories produces acceptable performance, let's keep it. It's easier to parse by a human.

**Review by hcho3 — APPROVED** [review]

**Pull request merged**